### PR TITLE
Add paint bucket module

### LINKED ISSUE #541: paint bucket (flood-fill) module
Breaking out of #165, this would take inputs of:

* a color to fill with
* x, y coordinate **starting pixel** to begin filling at
* (not in first version) - threshold -- see below

It would paint bucket fill starting at the given pixel and expanding outward.

Threshold means that the fill would fill any adjacent pixel of the same color OR within `threshold` value of that starting pixel color. It's expressed as a percentage difference from the original pixel color.

### PR DESCRIPTION
**Pull request by Mridul97**
Fixes #541 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts
* [ ] PR is descriptively titled
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by Mridul97** [comment]
@jywarren I tried to implement paint bucket module using flood-fill npm but some error came up, Could you please have a look at the code and help me to solve the error! Thanks!

**Comment by Mridul97** [comment]
I am not sure that is this the desired result, screenshot:

ss_30

Please have a look!

**Comment by Mridul97** [comment]
Yes, I will do that!

**Comment by Mridul97** [comment]
I have made the required changes. Please have a look!

Screenshot : 
ss_31

**Comment by HarshKhandeparkar** [comment]
@Mridul97 maybe sliders for rgb values will be better?

**Comment by Mridul97** [comment]
Yes, we can do that but I think many times users have exact rgb values of the color they want and may just want to copy it!

**Comment by HarshKhandeparkar** [comment]
@Mridul97 you are right on that one but can you add an optional alpha value?

**Comment by Mridul97** [comment]
@HarshKhandeparkar yes I will do that! Thanks!

**Comment by Mridul97** [comment]
I have made the changes. Please have a look! 
screenshot : 

ss_33

**Comment by Mridul97** [comment]
Done!

**Comment by jywarren** [comment]
Oh cool! But i'm not sure if the module is working properly - does it fill any polygon? Like this?

image


Maybe it just needs to be configured correctly? But this is very exciting -- thank you !!!!

**Comment by Mridul97** [comment]
@jywarren How a user will specify a polygon using x, y coordinates of starting pixel to begin filling at and a color to fill with. I mean like in the above flower pic if a user enters starting pixel, how would we know when to stop filling?

**Comment by jywarren** [comment]
I believe this is known as the "threshold" -- the simplest form would just
fill only pixels that neighbor and have identical color, but a "threshold"
implementation (we could do in follow-up) would allow filling into similar
colors, as determined by the threshold (usually expressed as a percentage!)

**Comment by Mridul97** [comment]
Oh! got it. Thanks!

**Comment by jywarren** [comment]
ah interesting! Can you try a crisper image maybe that has no anti-aliasing?

**Comment by Mridul97** [comment]
It seems to work similarly on aliasing and anti-aliasing images. The left image is with aliasing and right one is with anti-aliasing. Please have a look! 

screenshot 110

screenshot 111

screenshot 112

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "data-uri-to-buffer": "^2.0.0",
     "downloadjs": "^1.4.7",
     "fisheyegl": "^0.1.2",
+    "flood-fill": "^0.1.1",
```
[the remark this answers is not in the document]
Comment (Mridul97) [reply package.json]: No, The module didn't work for this, it looks like the module works only for black and white images.

**Comment by HarshKhandeparkar** [comment]
Maybe we can add tollerance instead of using `===` directly?
```js
if (pixels.get('same params', 0) > (r*(1 - tolerance/100)) && pixels.get('same params', 0) < (r*(1 + tolerance/100))){
  
}
```

Tollerance can be a percentage we can set by experimentation?

**Comment by Mridul97** [comment]
> Maybe we can add tollerance instead of using `===` directly?
> 
> ```js
> if (pixels.get('same params', 0) > (r*(1 - tolerance/100)) && pixels.get('same params', 0) < (r*(1 + tolerance/100))){
>   
> }
> ```
> 
> Tollerance can be a percentage we can set by experimentation?

Nice! I think this is same as 'Threshold' that @jywarren talked about above, maybe we can start by keeping it 5%?

**Comment by HarshKhandeparkar** [comment]
I think testing it at 5% will be good. Also making a new function for testing that might be good.
Maybe
```js
function isSimilarPixel(tolerance, other_params){
return bool;
}
```

**Comment by Mridul97** [comment]
> I think testing it at 5% will be good. Also making a new function for testing that might be good.
> Maybe
> 
> ```js
> function isSimilarPixel(tolerance, other_params){
> return bool;
> }
> ```

Yes, I was also thinking to do that! Thanks!

**Comment by Mridul97** [comment]
After adding threshold of 5%, @jywarren Please have a look! 

ss_46

**Comment by jywarren** [comment]
🎉🎉🎉🎉🎉🎉

Amazing. Is this ready to merge?

**Comment by Mridul97** [comment]
@jywarren Yes, I will rebase the branch then It is ready to be merged! :)

**Comment by HarshKhandeparkar** [comment]
@Mridul97 this is amazing! But does this work with other images maybe with a bit more noise?

**Comment by Mridul97** [comment]
> @Mridul97 this is amazing! But does this work with other images maybe with a bit more noise?

I would love to test it, do you have any such image on which we can try this out? Thanks!

**Comment by HarshKhandeparkar** [comment]
I don't know if this is a good example

images 1.

Can the module fill in one of those polygons?

**Comment by HarshKhandeparkar** [comment]
Also what will happen if there is a dot in between?


3wrxj

**Comment by Mridul97** [comment]
I got these results! 

ss_47

ss_48

**Comment by HarshKhandeparkar** [comment]
> I got these results!
> 
> ss_47
> 
> ss_48

The second one looka awesome but what happened to the first? Its pretty messy. Maybe the tolerance has to be set dynamically? Maybe we can do that in a followup?

**Comment by Mridul97** [comment]
Sounds good! any idea how will we set tolerance dynamically maybe take from the user?

**Comment by HarshKhandeparkar** [comment]
I don't really know but maybe we can somehow average the noise in lets say a rectangle around the starting pixel?

**Comment by HarshKhandeparkar** [comment]
Maybe get the sum of r,g,b,a values of each pixel in the rectangle and average it and subtract the sum of r,g,b,a value of starting pixel. This seems really overkill though 😅😅

**Comment by Mridul97** [comment]
Even I don't really know what would be right approach for this, lets think upon it on a follow-up PR! Thanks!

**Comment by Mridul97** [comment]
@jywarren I have rebased the branch! :)

**Comment by HarshKhandeparkar** [comment]
Or maybe the fastest way would be to subtract the r,g,b,a values of starting pixel from r,g,b,a values of the pixel adjacent to it and get a percentage for each value(i.e. r,g,b,a) then average that percentage? This might give us inconsistent and inaccurate results though.

File: src/modules/PaintBucket/PaintBucket.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          a = pixels.get(x, y, 3),
+          queuex = [x],
+          queuey = [y],
+          curry, currx,
+          north,
+          south,
+          n,
+          tolerance = 5;
+
+      function isSimilar(currx, curry){
```
Comment (HarshKhandeparkar) [inline src/modules/PaintBucket/PaintBucket.js]: You can return the logic statement used for the if block itself here and eliminate the if block.

**Comment by Mridul97** [comment]
> Or maybe the fastest way would be to subtract the r,g,b,a values of starting pixel from r,g,b,a values of the pixel adjacent to it and get a percentage for each value(i.e. r,g,b,a) then average that percentage? This might give us inconsistent and inaccurate results though.

But adjacent pixel could be same also, should we really keep the tolerance depending upon an adjacent pixel? Lets wait to hear what @jywarren has to say on this!

File: src/modules/PaintBucket/PaintBucket.js (hunk 2 above)
Comment (HarshKhandeparkar) [reply src/modules/PaintBucket/PaintBucket.js]: You can do that inside the isSimilar function.

File: src/modules/PaintBucket/PaintBucket.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        if (pixels.get(currx, curry, 0) > r*(1 - tolerance/100) && pixels.get(currx, curry, 0) < r*(1 + tolerance/100) &&
+            pixels.get(currx, curry, 1) > g*(1 - tolerance/100) && pixels.get(currx, curry, 1) < g*(1 + tolerance/100) &&
+            pixels.get(currx, curry, 2) > b*(1 - tolerance/100) && pixels.get(currx, curry, 2) < b*(1 + tolerance/100) &&
```
Comment (HarshKhandeparkar) [inline src/modules/PaintBucket/PaintBucket.js]: Also maybe you can assign 1-tolerance/100 and 1+tolerance/100 to different vars?

**Comment by jywarren** [comment]
Many graphics programs like Photoshop just have a default 5-15% tolerance,
but have it user-settable. I think we could expose one more user input and
use it as a parameter?

**Comment by Mridul97** [comment]
@jywarren Great! We can make it a range input where user can enter values between 5 and 15 for tolerance! Should we add this here or maybe in a follow-up pr? :)

File: src/modules/PaintBucket/PaintBucket.js (hunk 2 above)
Comment (Mridul97) [reply src/modules/PaintBucket/PaintBucket.js]: I have made the changes. Thanks!

File: src/modules/PaintBucket/PaintBucket.js (hunk 3 above)
Comment (Mridul97) [reply src/modules/PaintBucket/PaintBucket.js]: I have made the changes. Thanks!

**Comment by Mridul97** [comment]
@jywarren I have added tolerance as an user input, Please have a look! 

ss

**Comment by jywarren** [comment]
This looks fantastic. I'd like to request two follow up issues be opened: one to make the threshold range from 0-100, and another to propose an input type called something like "pixel" which prompts the user to click on the image and returns an xy coordinate. What do you think? We could use a crosshairs icon on a button and have a label like "Paint bucket starting point: click to select a pixel"

**Comment by jywarren** [comment]
Great work on this! It might be an interesting one to write a "looks-like" test for too - a circle before and after filling could make a good comparison!

**Pull request merged**